### PR TITLE
new syntax for lvalue references: `var b {.byaddr.} = expr`

### PR DESCRIPTION
**Pull request by timotheecour**
## TLDR
* introduces a new syntax for lvalue references: `var b {.byaddr.} = expr`, eg:
```nim
var b {.byaddr.} = foo[0].funWithSideEffect(2).bar.myStr
# simply use b, it'll be via reference semantically
b.add "foo"
```

* exact analog of C++ references `auto& b = expr;`
* much better approach than existing alternatives:
```nim
## existing alternative 1: template
template a: untyped = expr
  # * will evaluate expr each time it's used (even if costly, or has side effects)
  # * won't detect invalid `expr` if `a` is unused
  # * doesn't give redefinition errors

## existing alternative 2: let
let a = expr # not a reference, does a copy

## existing alternative 3: explicit `addr`
let a = expr.addr
a[]+=2 # less convenient: requires a deref on each use
fun(a) # less safe: the reference can escape
```

* supersedes #13342 which can be closed if this is merged
* supersedes #11824 which can be closed if this is merged
* see #11824 for rationale, examples etc

## compared to #13342 and #11824
(some of these are comparisons to only #13342 or only  #11824 or both)
* no parser change
* gives clean redefinition error despite https://github.com/nim-lang/Nim/issues/8275
in particular giving a redefinition error with #11824 approach would've been trickier to do correctly (we should only check in current scope)
```nim
# using #11824 approach:
import sugar
var x = @[1,2,3]
byAddr: x1=x[1]
x1+=10
byAddr: x1=x[2] # should give a redifinition error but doesn't, and hard to do (`when declared` is too naive)
```
* allows specifying type optionally:
`var b {.byaddr.}: int = s[0]`
* doesn't introduce a new keyword, only a new pragma just for `var` for now; so it's easier to extend, eg `var b {.byunsafeaddr.}: int = s[0]` if needed
* syntax is just a pragma extension of a regular let statement
* addresses the essential technical criticism from https://github.com/nim-lang/Nim/pull/13342#issuecomment-582835732
> And personally I would have used let/var x {.view.} = someLocation. Adding a new pragma before a new keyword seems good and so does keeping the let vs var distinction.

I can change to `view` which is short and explicit, but I'd rather do the change just once, after this is greenlighted

## example
```nim
    var s = @[10,11,12]
    var a {.byaddr.} = s[0]
    a+=100
    doAssert s == @[110,11,12]
    doAssert a is int
    var b {.byaddr.}: int = s[0]
    doAssert a.addr == b.addr
```

## future PR's
* for now the var section with `byaddr` must have a single element, support for more than 1 could be added in future (will be backward compatible of course)
```nim
# ok
var
  b {.byaddr.} = expr

# not implemented
var
  b1 = expr1
  b2 {.byaddr.} = expr2
  b3 {.byaddr.} = expr2
```

* we could introduce `let b {.byaddr.} = expr` if needed
* tuple ref-unpacking as proposed here https://github.com/nim-lang/RFCs/issues/194#issuecomment-591629128:
```
var x = (1, 2)
var (a,b) {.byaddr.} = x
```

## note
this should replace at least some use cases for `shallowCopy` and `{.shallow.} `

### CODE REVIEW

**Comment by Araq** [comment]
Pretty good. However, either use an AST to AST transformation inside semexprs only without the helper public symbols in system.nim, or let's fix pragmas for variables so that is can be done entirely as a template/macro.

**Comment by kaushalmodi** [comment]
@timotheecour Can you please also document this in the manual in the Pragmas section: https://nim-lang.github.io/Nim/manual#pragmas ? Thanks.

**Comment by krux02** [comment]
Aren't references a feature that is left out in Nim for reasons?

**Comment by timotheecour** [comment]
@araq PTAL
> or let's fix pragmas for variables so that is can be done entirely as a template/macro.

done

compiler now lowers `var a {.foo.}: MyType = expr` to `foo(a, MyType, expr)` with user defined `foo` (with appropriate checks)

there are othe applications for this new compiler feature besides the current `byaddr`, eg:
```
# wraps a decl behind getter and setter accessors, eg for debugging
var {.onAccessCallback.} x: int 
```

**Comment by timotheecour** [comment]
@Araq friendly ping (I'll rebase to fix `changelog.md` conflict after you greenlight this as the changelog changes often)

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -431,7 +431,53 @@ proc setVarType(c: PContext; v: PSym, typ: PType) =
...
+    let sym = searchInScopes(c, ident)
+    if sfCustomPragma in sym.flags: return nil # skip `template myAttr() {.pragma.}`
+    let lhs = b[0]
+    let clash = strTableGet(c.currentScope.symbols, lhs.ident)
+    if clash != nil:
+      # refs https://github.com/nim-lang/Nim/issues/8275
+      wrongRedefinition(c, lhs.info, lhs.ident.s, clash.info)
+
+    result = newTree(nkCall)
+    doAssert nodePragma.kind == nkIdent, $nodePragma.kind
```
Comment (Araq) [inline compiler/semstmts.nim]: But previously ``nodePragma`` could also be a ``nkAccQuoted``!

File: compiler/semstmts.nim [hunk 2]
```diff
@@ -431,7 +431,53 @@ proc setVarType(c: PContext; v: PSym, typ: PType) =
...
 
+proc semLowerLetVarCustomPragma(c: PContext, a: PNode, n: PNode): PNode =
+  var b = a[0]
+  if b.kind == nkPragmaExpr:
+    if b[1].len != 1:
+      # we could in future support pragmas w args eg: `var foo {.bar:"goo".} = expr`
+      return nil
+    let nodePragma = b[1][0]
+    # see: `singlePragma`
+    if nodePragma.kind notin {nkIdent, nkAccQuoted}:
```
Comment (Araq) [inline compiler/semstmts.nim:442]: I think there can also be ``nkOpenSymChoice``, ``nkClosedSymChoice``, see ``considerQuotedIdent``.

File: compiler/semstmts.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/semstmts.nim]: done, and added a test

File: compiler/semstmts.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/semstmts.nim:442]: please provide an example that would require this; in any case, this could be added in future PR's as future enhancement (more things (not less) becoming possible with var/let user pragmas) without breaking code.
